### PR TITLE
add updateHref to templateUtility

### PR DESCRIPTION
**Pull request by superyyrrzz**
usage:
```js
model.href = templateUtility.getHrefFromRoot(model.href, model._key);
```

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs [hunk 1]
```diff
@@ -38,5 +38,20 @@ public string ResolveSourceRelativePath(string originPath, string currentFileOut
                 return originPath;
             }
         }
+
+        public string UpdateHref(string originalHref, string sourceFileKey)
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs]: >UpdateHref [](start = 22, length = 10)

is it possible to extract and reuse DocumentBuildContext.UpdateHrefCore?

Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs]: a better naming? "update" sounds obscure

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs]: It's different. DocumentBuildContext.UpdateHrefCore converts href starting with `~/` to the final href, while this PR's function converts original href to href starting with `~/`.

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs]: get it, it's better to have better naming.

---
In reply to: [110823857](https://github.com/dotnet/docfx/pull/1532#discussion_r110823857) [](ancestors = 110823857)

File: src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs [hunk 2]
```diff
@@ -38,5 +38,20 @@ public string ResolveSourceRelativePath(string originPath, string currentFileOut
...
+        {
+            if (string.IsNullOrEmpty(sourceFileKey) || string.IsNullOrEmpty(originalHref) || !RelativePath.IsRelativePath(originalHref))
+            {
+                return originalHref;
+            }
+            var path = (RelativePath)sourceFileKey + (RelativePath)UriUtility.GetPath(originalHref);
+            var file = path.GetPathFromWorkingFolder().UrlDecode();
+            if (!_context.AllSourceFiles.ContainsKey(file))
+            {
+                return originalHref;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs]: >originalHref [](start = 23, length = 12)

in what case the source files not contains the file? need we log warning here?

Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Build.Engine/TemplatePreprocessors/TemplateUtility.cs]: yes it makes sense. resolved

**Comment by hellosnow** [comment]
:shipit:

**Pull request merged**